File: db-async-common/src/main/scala/com/github/mauricio/async/db/pool/PartitionedAsyncObjectPool.scala
package com.github.mauricio.async.db.pool

import scala.concurrent.Future
import com.github.mauricio.async.db.util.ExecutorServiceUtils
import scala.concurrent.Promise
import java.util.concurrent.ConcurrentHashMap
import scala.util.Success
import scala.util.Failure

class PartitionedAsyncObjectPool[T](
    factory: ObjectFactory[T],
    configuration: PoolConfiguration,
    numberOfPartitions: Int)
    extends AsyncObjectPool[T] {

    import ExecutorServiceUtils.CachedExecutionContext

    private val pools =
        (0 until numberOfPartitions)
            .map(_ -> new SingleThreadedAsyncObjectPool(factory, partitionConfig))
            .toMap

    private val checkouts = new ConcurrentHashMap[T, SingleThreadedAsyncObjectPool[T]]

    def take: Future[T] = {
        val pool = currentPool
        pool.take.andThen {
            case Success(conn) =>
                checkouts.put(conn, pool)
            case Failure(_) =>
        }
    }

    def giveBack(item: T) =
        checkouts
            .remove(item)
            .giveBack(item)
            .map(_ => this)

    def close =
        Future.sequence(pools.values.map(_.close)).map {
            _ => this
        }

    def availables: Traversable[T] = pools.values.map(_.availables).flatten

    def inUse: Traversable[T] = pools.values.map(_.inUse).flatten

    def queued: Traversable[Promise[T]] = pools.values.map(_.queued).flatten

    protected def isClosed =
        pools.values.forall(_.isClosed)

    private def currentPool =
        pools(currentThreadAffinity)

    private def currentThreadAffinity =
        (Thread.currentThread.getId % numberOfPartitions).toInt

    private def partitionConfig =
        configuration.copy(
            maxObjects = configuration.maxObjects / numberOfPartitions,
            maxQueueSize = configuration.maxQueueSize / numberOfPartitions
        )
}

File: db-async-common/src/main/java/com/github/jasync/sql/db/pool/PartitionedAsyncObjectPool.kt
package com.github.jasync.sql.db.pool

import com.github.jasync.sql.db.util.ExecutorServiceUtils
import com.github.jasync.sql.db.util.mapAsync
import java.util.concurrent.CompletableFuture
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.Executor


open class PartitionedAsyncObjectPool<T>(
    factory: ObjectFactory<T>,
    private val configuration: PoolConfiguration,
    private val numberOfPartitions: Int,
    private val executionContext: Executor = ExecutorServiceUtils.CommonPool)
  : AsyncObjectPool<T> {

  private val pools: Map<Int, SingleThreadedAsyncObjectPool<T>> =
      (0 until numberOfPartitions)
          .map { it to SingleThreadedAsyncObjectPool(factory, partitionConfig()) }.toMap()


  private val checkouts = ConcurrentHashMap<T, SingleThreadedAsyncObjectPool<T>>()

  override fun take(): CompletableFuture<T> {
    val pool = currentPool()
    return pool.take().mapAsync(executionContext) {
      checkouts[it] = pool
      it
    }
  }

  override fun giveBack(item: T): CompletableFuture<AsyncObjectPool<T>> {
    val removed = checkouts.remove(item)!!
    val singleRemoved = removed.giveBack(item)
    return singleRemoved.mapAsync(executionContext) { this }
  }

  override fun close(): CompletableFuture<AsyncObjectPool<T>> =
      CompletableFuture.allOf(* pools.values.map { it.close() }.toTypedArray()).mapAsync(executionContext) { this }

  fun availables(): List<T> = pools.values.map { it.availables() }.flatten()

  fun inUse(): List<T> = pools.values.map { it.inUse() }.flatten()

  fun queued(): List<CompletableFuture<T>> = pools.values.map { it.queued() }.flatten()

  protected fun isClosed() =
      pools.values.all { it.isClosed() }

  private fun currentPool() = pools.getValue(currentThreadAffinity())

  private fun currentThreadAffinity() = (Thread.currentThread().getId() % numberOfPartitions).toInt()

  private fun partitionConfig() =
      configuration.copy(
          maxObjects = configuration.maxObjects / numberOfPartitions,
          maxQueueSize = configuration.maxQueueSize / numberOfPartitions
      )
}
